includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -276,6 +276,79 @@ public function sanitize() {
+	 * @return array Augmented rule spec.
+	 */
+	private function get_rule_spec_list_to_validate( $node, $rule_spec ) {
+
+		// Expand extension_spec into a set of attr_spec_list.
+		if ( isset( $rule_spec[ AMP_Rule_Spec::TAG_SPEC ]['extension_spec'] ) ) {
+			$extension_spec = $rule_spec[ AMP_Rule_Spec::TAG_SPEC ]['extension_spec'];
+			$custom_attr    = 'amp-mustache' === $extension_spec['name'] ? 'custom-template' : 'custom-element';
+
+			$rule_spec[ AMP_Rule_Spec::ATTR_SPEC_LIST ][ $custom_attr ] = array(
+				AMP_Rule_Spec::VALUE     => $extension_spec['name'],
+				AMP_Rule_Spec::MANDATORY => true,
+			);
+
+			$versions = array_unique( array_merge(
+				isset( $extension_spec['allowed_versions'] ) ? $extension_spec['allowed_versions'] : array(),
+				isset( $extension_spec['version'] ) ? $extension_spec['version'] : array()
+			) );
+
+			$rule_spec[ AMP_Rule_Spec::ATTR_SPEC_LIST ]['src'] = array(
+				AMP_Rule_Spec::VALUE_REGEX => implode( '', array(
+					'^',
+					preg_quote( 'https://cdn.ampproject.org/v0/' . $extension_spec['name'] . '-' ),
+					'(' . implode( '|', $versions ) . ')',
+					'\.js$',
+				) ),
+			);
+		}
+
+		// Augment the attribute list according to the parent's reference points, if it has them.
+		if ( ! empty( $node->parentNode ) && isset( $this->allowed_tags[ $node->parentNode->nodeName ] ) ) {
+			foreach ( $this->allowed_tags[ $node->parentNode->nodeName ] as $parent_rule_spec ) {
+				if ( empty( $parent_rule_spec[ AMP_Rule_Spec::TAG_SPEC ]['reference_points'] ) ) {
+					continue;
+				}
+				foreach ( $parent_rule_spec[ AMP_Rule_Spec::TAG_SPEC ]['reference_points'] as $reference_point_spec_name => $reference_point_spec_instance_attrs ) {
+					$reference_point = AMP_Allowed_Tags_Generated::get_reference_point_spec( $reference_point_spec_name );
+					if ( empty( $reference_point[ AMP_Rule_Spec::ATTR_SPEC_LIST ] ) ) {
+						/*
+						 * See special case for amp-selector in AMP_Tag_And_Attribute_Sanitizer::is_amp_allowed_attribute()
+						 * where its reference point applies to any descendant elements, not just direct children.
+						 */
+						continue;
+					}
+					foreach ( $reference_point[ AMP_Rule_Spec::ATTR_SPEC_LIST ] as $attr_name => $reference_point_spec_attr ) {
+						$reference_point_spec_attr = array_merge(
+							$reference_point_spec_attr,
+							$reference_point_spec_instance_attrs
+						);
+
+						/*
+						 * Ignore mandatory constraint for now since this would end up causing other sibling children
+						 * getting removed due to missing a mandatory attribute. To sanitize this it would require
+						 * higher-level processing to look at an element's surrounding context, similar to how the
+						 * sanitizer does not yet handle the mandatory_oneof constraint.
+						 */
+						unset( $reference_point_spec_attr['mandatory'] );
+
+						$rule_spec[ AMP_Rule_Spec::ATTR_SPEC_LIST ][ $attr_name ] = $reference_point_spec_attr;
+					}
+				}
+			}
+		}
+		return $rule_spec;
```

Comment:
Most of this code is refactored into a separate method from below. See https://github.com/Automattic/amp-wp/pull/1386/commits/002bda35034eca01febaf681906ed9624059bf31 for the actual changes introduced.